### PR TITLE
[runner] add stats only for checks currently in the schedule

### PR DESCRIPTION
**Pull request by truthbk**
### What does this PR do?

Supersedes #2301 - but with a cleaner approach. Since the first thing we do when stopping a check is remove it from the schedule, if it runs after that, then it should suffice with just checking with the scheduler if that check should still be considered. By doing so, we avoid the need to keep track of stopped checks, if it's not in the scheduler, it's been stopped. 

### Motivation

Inconsistencies in reporting, execution in the runner due to being oblivious to the state of the scheduler.

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/collector/py/check.go [hunk 1]
```diff
@@ -50,12 +52,19 @@ func NewPythonCheck(name string, class *python.PyObject) *PythonCheck {
 		interval:     check.DefaultCheckInterval,
 		lastWarnings: []error{},
 	}
+	pyCheck.stopped.Store(false)
```
Comment (olivielpeau) [inline pkg/collector/py/check.go]: With the solution in this PR, I don't think the changes here (`pkg/collector/py/check.go`) really make sense. These checks should be common to all types of checks IMO, and should be done from the runner before calling `check.Run()`

File: pkg/collector/runner/runner.go [hunk 2]
```diff
@@ -205,12 +214,14 @@ func (r *Runner) StopCheck(id check.ID) error {
 	defer r.m.Unlock()
 
 	if c, isRunning := r.runningChecks[id]; isRunning {
-		log.Debugf("Stopping check %s", c)
+		log.Debugf("Stopping check %s %s", string(id), c)
```
Comment (olivielpeau) [inline pkg/collector/runner/runner.go]: I think we should log `c.ID()` instead, it should be human-readable

File: pkg/collector/runner/runner.go [hunk 3]
```diff
@@ -205,12 +214,14 @@ func (r *Runner) StopCheck(id check.ID) error {
...
 		go func() {
+			// Remember that the check was stopped so that even if it runs we can discard its stats
 			c.Stop()
 			close(done)
 		}()
 	} else {
+		log.Debugf("check %s is not running", string(id))
```
Comment (olivielpeau) [inline pkg/collector/runner/runner.go]: nit: capitalize first letter, clarify message (maybe add `not stopping it` or something similar), and use `c.ID()`

File: pkg/collector/scheduler/scheduler.go [hunk 4]
```diff
@@ -99,7 +99,7 @@ func (s *Scheduler) Cancel(id check.ID) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	log.Infof("Unscheduling check %v ", check.IDToCheckName(id))
+	log.Infof("Unscheduling %s check %v ", string(id), check.IDToCheckName(id))
```
Comment (olivielpeau) [inline pkg/collector/scheduler/scheduler.go]: let's just log the `id` if it's human-readable?

**Review by olivielpeau — COMMENTED** [review]
LGTM for 6.5 👌 left a few nits.

I'm thinking https://github.com/DataDog/datadog-agent/pull/2301#pullrequestreview-153754022 could be a nicer long-term approach to the problem (could avoid the dependency of the runner on the scheduler), maybe as an improvement for later.

File: pkg/collector/py/check.go (hunk 1 above)
Comment (truthbk) [reply pkg/collector/py/check.go]: This should be fine, it was a way to at least get rid of the error, given the numerous race conditions that were getting sorted with the other PRs.

**Review by olivielpeau — DISMISSED** [review]

File: pkg/collector/runner/runner.go [hunk 5]
```diff
@@ -58,8 +59,8 @@ type runnerCheckStats struct {
 // Runner ...
 type Runner struct {
 	pending          chan check.Check         // The channel where checks come from
-	done             chan bool                // Guard for the main loop
```
Comment (olivielpeau) [inline pkg/collector/runner/runner.go]: just curious: why did you remove this?

File: pkg/collector/scheduler/scheduler.go [hunk 6]
```diff
@@ -110,6 +110,7 @@ func (s *Scheduler) Cancel(id check.ID) error {
 	if err != nil {
 		return fmt.Errorf("unable to remove the Job from the queue: %s", err)
 	}
+	delete(s.checkToQueue, id)
```
Comment (olivielpeau) [inline pkg/collector/scheduler/scheduler.go]: nice catch 👌 

Originally an oversight from https://github.com/DataDog/datadog-agent/pull/2127, but at that point only caused a very small mem leak in very dynamic AD environments. With the new logic in this PR, this oversight caused a new problem now that we use this map to determine if a check is still in the schedule (in `IsCheckScheduled`)

**Review by olivielpeau — APPROVED** [review]
LGTM, added a couple of comments for clarity (feel free to merge after answering to the first comment, thanks!)

File: pkg/collector/runner/runner.go (hunk 5 above)
Comment (truthbk) [reply pkg/collector/runner/runner.go]: Because we don't use it anywhere.
